lib/proxy/pluggable.rb

```diff
@@ -100,6 +86,7 @@ def default_settings(a_hash = {})
     end
 
     def initialize_after(*module_names)
+      logger.warn("#{plugin_name}: 'initialize_after' has been deprecated and will be removed in future versions.")
```

Comment:
> Was there any use for feature this in existing plugins?

This feature is not being used In plugins I checked (all under foreman org in github). 

> It looks like we now lose the ability of a provider or module to order itself after another module, and I don't know if we even take the require into account for ordering. It still seems useful to me, despite the general ordering improvements in this PR.

The goal is to remove guarantees about the order in which modules are loaded (modules are supposed to be independent of each other), other than providers before main modules. The functionality was introduced before providers, with idea was that plugins will be used to implement provider logic. Lack of use in real world in combination with not being able to handle dependency loops makes me want to remove the functionality rather than fix it.

Please note that the less strict module dependency functionality  is present and working (probably should be fixed to take into account enabled modules, as opposed to loaded ones).

> It's not so much deprecated as disabled, so the code below should be removed as I don't think it's functional.
> If being removed, it would be better to raise an exception as it's going to behave very differently.
> 

True, will remove the code and raise an exception instead.